Add "comment-pattern" rule

omril321 opened the pull request:
<!-- Each pull request must be associated with an open issue unless it's a documentation fix. If a corresponding issue does not exist, please create one so we can discuss the change first. -->

<!-- Please answer the following. We close pull requests that don't. -->

> Which issue, if any, is this issue related to?

Closes [#4887](https://github.com/stylelint/stylelint/issues/4887)

> Is there anything in the PR that needs further explanation?

No, it's self-explanatory.

I do want to mention that it is my first PR for an open source project. I'll appreciate any input :)

hudochenkov commented on docs/user-guide/rules/list.md:
```diff
@@ -230,6 +230,7 @@ Grouped first by the following categories and then by the [_thing_](http://apps.
 
 - [`comment-word-blacklist`](../../../lib/rules/comment-word-blacklist/README.md): Specify a list of disallowed words within comments. **(deprecated)**
 - [`comment-word-disallowed-list`](../../../lib/rules/comment-word-disallowed-list/README.md): Specify a list of disallowed words within comments.
+- [`comment-pattern`](../../../lib/rules/comment-pattern/README.md): Specify a pattern for comments.
```
Could you put it in the beginning of the list, please? We sort all rules in a section alphabetically.

hudochenkov commented on lib/rules/comment-pattern/README.md:
````diff
@@ -0,0 +1,39 @@
+
+<!-- prettier-ignore -->
+```css
+                  /*comment*/
+                 /* ↑ */ a { --foo-: 1px; }
+/**                 ↑
+ * The pattern of this */
````
Simplify and leave only related CSS.

```suggestion
/*  comment */
/** ↑
 * The pattern of this */
```

hudochenkov commented on lib/rules/comment-pattern/README.md:
````diff
@@ -0,0 +1,39 @@
+
+## Options
+
+`regex`
+
+The pattern to match the comments with.
+
+Given the regex:
+
+```js
+/foo .+/
+```
````
I would use [this](https://github.com/stylelint/stylelint/blame/924646681a355b327ccfeb6f00be4e1636333911/lib/rules/custom-property-pattern/README.md#L14-L22) and only adjusted RegEx.

If rule accept RegEx, then we allow it to be both string and RegEx, because JSON and other formats doesn't have RegEx as JavaScript does.

hudochenkov commented on lib/rules/comment-pattern/__tests__/index.js:
```diff
@@ -0,0 +1,84 @@
+			code: '// foo-ok',
+			description: 'ignored inline scss comments',
+		},
+		{
+			code: '.without-comment { }',
+			description: 'ignores scss without comments',
+		},
```
This test is not necessary, as it's exactly the same as for standard syntax.

hudochenkov commented on lib/rules/comment-pattern/__tests__/index.js:
```diff
@@ -0,0 +1,84 @@
+	accept: [
+		{
+			code: 'a {} // foo-ok',
+			description: 'ignored inline scss comments',
+		},
+		{
+			code: '// foo-ok',
+			description: 'ignored inline scss comments',
```
I suggest remove these descriptions, because double-slash comment are not ignored.

hudochenkov commented on lib/rules/comment-pattern/__tests__/index.js:
```diff
@@ -0,0 +1,84 @@
+
+testRule({
+	ruleName,
+	config: [/foo-.+/],
```
Could you also make a copy of this `testRule`, but use `'foo-.+'` string instead of RegEx. To make sure both RegEx and string option work exactly the same.

omril321 replied:
Instead of copy-pasting the entire call for `testRule`, what do you think about iterating the configurations that should behave in the exact same way, and calling `testRule` for each?
Meaning:
```js
[/foo-.+/, 'foo-.+'].forEach((config) => {
	testRule({
		ruleName,
		config: [config],
                 ...
```

This way, it is clear that the two configurations behave exactly the same, and if for some reason the tests will require updates - it will be in a single place.
The downside that it could be less readable when seeing for the first time (I didn't find something similar in the existing tests)

hudochenkov commented on lib/rules/comment-pattern/index.js:
```diff
@@ -0,0 +1,47 @@
+const ruleName = 'comment-pattern';
+
+const messages = ruleMessages(ruleName, {
+	expected: 'Expected comment to match specified pattern',
```
I think message would be more understandable if we show pattern in it:

```
Expected comment to match specified pattern: /foo.+/
```

What do you think, @stylelint/contributors?

hudochenkov commented on lib/rules/comment-pattern/index.js:
```diff
@@ -0,0 +1,47 @@
+
+function rule(pattern) {
+	return (root, result) => {
+		const validOptions = validateOptions(result, ruleName, {
+			actual: pattern,
+			possible: [_.isRegExp],
+		});
+
+		if (!validOptions) {
+			return;
+		}
```
It needs to support string as I described in the other comment. You can use this code: https://github.com/stylelint/stylelint/blob/924646681a355b327ccfeb6f00be4e1636333911/lib/rules/custom-property-pattern/index.js#L19-L28

hudochenkov reviewed (changes requested):
That's a great first pull request!

Few small changes required.

CI is failing: To fix formatting errors run `npm run format`. To run all test locally use `npm test`, if will ensure all tests are passing and not lint errors.

omril321 commented:
Thanks for the feedback! I've implemented the fixes as suggested. Please let me know if further changes are required.
By the way, as a feedback of my own, I didn't know that adding a rule which accepts `regex` as an option requires support for a string option as well. I think I got confused by [the original message here](https://github.com/stylelint/stylelint/issues/4887#issuecomment-683840236)
> ...
> Primary option: regex
> ...

I understand now why a string option needs to be supported as well, but I believe that specifying it in the requirements will make it more obvious for first-timers :)

hudochenkov commented:
> By the way, as a feedback of my own, I didn't know that adding a rule which accepts regex as an option requires support for a string option as well. I think I got confused by the original message here

Indeed, it was missed. I've added it to mentioned message. Thanks!

hudochenkov commented on lib/rules/comment-pattern/__tests__/index.js:
```diff
@@ -28,11 +28,46 @@ testRule({
+	reject: [
+		{
+			code: '/* not foo- */',
+			message: messages.expected(/foo-.+/),
+		},
+		{
+			code: '/**/',
+			message: messages.expected(/foo-.+/),
```
It should match config, and in config we have string.

```suggestion
			message: messages.expected('foo-.+'),
		},
		{
			code: '/**/',
			message: messages.expected('foo-.+'),
```

omril321 replied:
@hudochenkov got it, fixed now. 
I thought that we wanted to show the `normalizedPattern` in the message, instead of the pattern that was supplied in the config. I've updated the implementation + tests as suggested :)

hudochenkov replied:
> I thought that we wanted to show the normalizedPattern in the message, instead of the pattern that was supplied in the config.

I didn't thought about that :) Now I'm not sure which one is better :)

@stylelint/contributors what do you think?

ybiquitous replied:
It seems reasonable to me to always show the same form `pattern "foo"`, not `pattern "/foo/"`.
Even in different config file formats, it seems users expect the same output of a message, for example:

```yaml
# .stylelintrc.yml
rules:
  comment-pattern: ["foo"]
```

```js
// .stylelintrc.js
module.exports = {
  rules: {
    "comment-pattern": [/foo/] // equal to ["foo"]
  }
}
```

```console
$ bin/stylelint.js a.css --config .stylelintrc.yml
a.css
 1:1  ✖  Expected comment to match specified pattern "foo"   comment-pattern

$ bin/stylelint.js a.css --config .stylelintrc.js
a.css
 1:1  ✖  Expected comment to match specified pattern "/foo/"   comment-pattern
```

omril321 replied:
@ybiquitous forgive me but I might be confused.. Are you suggesting to strip the slashes from the pattern when generating an error message? If so, how should regexes with modifiers look? e.g. `/foo-.+/i` or `/foo-.+/ig`?

ybiquitous replied:
> Are you suggesting to strip the slashes from the pattern when generating an error message?

Yes, I thought we could keep the same appearance of an error message in any config formats.

> If so, how should regexes with modifiers look? e.g. `/foo-.+/i` or `/foo-.+/ig`?

Sorry, I missed regex modifiers. Surely your current implementaion seems better if we need to include the modifiers. 👍

hudochenkov reviewed (changes requested):
Thank for making changes. One small change to make.

jeddy3 commented on lib/rules/comment-pattern/__tests__/index.js:
```diff
@@ -0,0 +1,113 @@
+			\t
+				*/`,
+		},
+		{
```
We can remove these `.without-comment {}` tests as they are duplicated in the [basic checks](https://github.com/stylelint/stylelint/blob/8366b54f536b26c2a994865568336c577ea5625e/lib/testUtils/basicChecks.js).

jeddy3 commented on lib/rules/comment-pattern/index.js:
```diff
@@ -0,0 +1,48 @@
+const ruleName = 'comment-pattern';
+
+const messages = ruleMessages(ruleName, {
+	expected: (pattern) => `Expected comment to match specified pattern "${pattern}"`,
```
```suggestion
	expected: (pattern) => `Expected comment to match pattern "${pattern}"`,
```

Let's drop "specified" as we show the pattern in the message. I think including the pattern is a good idea, but let's create a follow-up issue to do this consistently across the other `*-pattern` rules.

omril321 replied:
Great. I can open an issue soon

jeddy3 commented on lib/rules/comment-pattern/__tests__/index.js:
```diff
@@ -0,0 +1,113 @@
+	],
+});
+
+testRule({
```
I think we can remove this `testRule`.

---

I'd expect `config: [true]` to produce an invalid option error. This is likely a short coming in the `validateOptions` function, that we can pick up another time.

omril321 replied:
I removed this. I'll open an issue soon 👍

jeddy3 reviewed (changes requested):
@omril321 Looking good, thank you. Just a couple of minor requests from me.

ybiquitous commented on lib/rules/comment-pattern/README.md:
```diff
@@ -0,0 +1,38 @@
+# custom-property-pattern
```
Oh, I noticed the typo. 😅 

```suggestion
# comment-pattern
```

omril321 replied:
well, that's embarrassing 😅

ybiquitous reviewed (approved):
LGTM! 👏

jeddy3 reviewed (changes requested):
This looks good to me, once the redundant `code: '.without-comment {}',` are removed.

omril321 commented:
> This looks good to me, once the redundant `code: '.without-comment {}',` are removed.

Thanks for your patience 🙏  I wasn't thorough enough when I removed redundant tests . I've just removed them

jeddy3 reviewed (approved):
Thanks for making the changes.

LGTM. Thank you for contributing to stylelint!

hudochenkov commented:
> - Added: `comment-pattern` rule ([#4962](https://github.com/stylelint/stylelint/pull/4962)).
